lib/serverenroll_test.go

```diff
@@ -170,3 +171,53 @@ func TestPasswordLimit(t *testing.T) {
 	util.FatalError(t, err, "Failed to get 'user1' from database")
 	assert.Equal(t, 0, user1.GetFailedLoginAttempts())
 }
+
+func TestCertificateExpiration(t *testing.T) {
+	cleanTestSlateSE(t)
+	defer cleanTestSlateSE(t)
+
+	srv := TestGetRootServer(t)
+	err := srv.Start()
+	util.FatalError(t, err, "Failed to start server")
+	defer srv.Stop()
+
+	caCertPem, _ := srv.CA.getCACert()
+	caCert, _ := util.GetX509CertificateFromPEM(caCertPem)
+
+	client := getTestClient(rootPort)
+
+	csrPEM, _, err := client.GenCSR(&api.CSRInfo{CN: "admin"}, "admin")
+	assert.NoError(t, err, "Failed to generate CSR")
+
+	reqNet := &api.EnrollmentRequestNet{
+		SignRequest: signer.SignRequest{
+			Request: string(csrPEM),
+			// requesting certificate with validity time wider then CA cert
+			NotBefore: caCert.NotBefore.Add(-1 * time.Hour),
+			NotAfter:  caCert.NotAfter.Add(24 * time.Hour),
+		},
+	}
+
+	body, err := util.Marshal(reqNet, "SignRequest")
+	assert.NoError(t, err, "Failed to marshal enroll request")
+
+	// Send the CSR to the fabric-ca server with basic auth header
+	post, err := client.newPost("enroll", body)
+	assert.NoError(t, err, "Failed to create post request")
+	post.SetBasicAuth("admin", "adminpw")
+
+	var result api.EnrollmentResponseNet
+	err = client.SendReq(post, &result)
+	assert.NoError(t, err, "Failed to enroll")
+
+	// verify response
+	certBytes, err := util.B64Decode(result.Cert)
+	assert.NoError(t, err, "Failed to convert certificate")
+	userCert, err := util.GetX509CertificateFromPEM(certBytes)
+	assert.NoError(t, err, "Failed to extract certificate from enroll response")
+
+	assert.False(t, userCert.NotBefore.Before(caCert.NotBefore),
+		"user certificate NotBefore %v is before CA cert NotBefore %v", userCert.NotBefore, caCert.NotBefore)
+	assert.False(t, userCert.NotAfter.After(caCert.NotAfter),
+		"user certificate NotAfter %v is after CA cert NotAfter %v", userCert.NotAfter, caCert.NotAfter)
```

Comment:
@darekzbik 

Thanks for taking this on.  Nothing wrong with what you've done, but a couple of suggestions:

While these checks are fine, shouldn't you actually check that NotBefore and NotAfter are equal to the CA cert values as well?

And should there also be a test where NotBefore and NotAfter from the enroll request are in the valid range?

Reply:
1.  I did not check equality.  I was focused on checking if created certificate dates are in range of CA cert validity dates. Checking equality is fine as long as we assume that CA server is creating certificates with validity period as wide as possible. EQ test will fail when CA server will be changed and certs created by CA will be time limited (let say no longer than 360 days from now).

1. In this PR I touch only ‘NotBefore’ field in the cert. Logic of ‘userCert.NotAfter’ was previously in the code, probably not tested in the same way. I was focused on my change only. I’ll odd some more check in the same test.

In fact I'm surprised that CA server is issuing certs with past dates when someone ask for it. I don't like such behaviour but such change has potentially bigger impact.

Reply:
@mastersingh24 plese see this https://github.com/hyperledger/fabric-ca/pull/206